proposals/nnnn-span-access-shared-contiguous-storage.md

```diff
@@ -0,0 +1,821 @@
+
+1. The pointer itself is unsafe and unmanaged
+2. `subscript` is only bounds-checked in debug builds of client code
+3. It might escape the duration of the closure
+
+Even if the body of the `withUnsafeXXX` call does not escape the pointer, other functions called within the closure have to be written in terms of unsafe pointers. This requires programmer vigilance across a project and potentially spreads the use of unsafe types, even if the helper functions could have been written in terms of safe constructs.
+
+## Proposed solution
+
+#### `Span`
+
+`Span` will allow sharing the contiguous internal representation of a type, by providing access to a borrowed view of an interval of contiguous memory. `Span` does not copy the underlying data: it instead relies on a guarantee that the original container cannot be modified or destroyed while the `Span` exists. In the prototype that accompanies this first proposal, `Span`s will be constrained to closures from which they structurally cannot escape. Later, we will introduce a lifetime dependency between a `Span` and the binding of the type vending it, preventing its escape from the scope where it is valid for use. Both of these approaches guarantee temporal safety. `Span` also performs bounds-checking on every access to preserve spatial safety. Additionally `Span` always represents initialized memory, preserving the definite initialization guarantee.
+
+`Span` is intended as the currency type for local processing over values in contiguous memory. It is a replacement for many API currently using `Array`, `UnsafeBufferPointer`, `Foundation.Data`, etc., that do not need to escape the owning container.
+
+A `Span` provided by a container represents a borrow of that container. `Span` can therefore provide simultaneous access to a non-copyable container. It can also help avoid unwanted copies of copyable containers. Note that `Span` is not a replacement for a copyable container with owned storage; see [future directions](#Directions) for more details ([Resizable, contiguously-stored, untyped collection in the standard library](#Bytes))
+
+#### `RawSpan`
+
+`RawSpan` allows sharing contiguous memory representing values which may be heterogeneously-typed, such as memory intended for parsing. It makes the same safety guarantees as `Span`. Since it is a fully concrete type, it can achieve great performance in debug builds of client code as well as straightforward performance in library code.
+
+A `RawSpan` can be obtained from containers of `BitwiseCopyable` elements, as well as be initialized directly from an instance of `Span<T: BitwiseCopyable>`.
+
+## <a name="Design"></a>Detailed design
+
+`Span<Element>` is a simple representation of a region of initialized memory.
+
+```swift
+@frozen
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: UnsafePointer<Element>
```

Comment:
`Span` has the wrong pointer type: `UnsafePointer<Element>`. The underlying pointer should be `UnsafeRawPointer`. An important goal of Span is that it can be used to reinterpret memory as different `BitwiseCopyable` elements. It needs to be possible to convert `Span<T: BitwiseCopyable>` to `Span<U: BitwiseCopyable>`. That should require an explicit unsafe API but should still be well-defined. If you do this conversion with an underlying typed pointer, you get undefined behavior. We want multiple read-only spans to safely view the same underlying memory as different types. We also want to be able to view slices of a `RawSpan` as a particular element type. This is the same problem. Note that `RawSpan` does not provide any additional functionality; it is only a convenience for buffers of heterogenous elements, where declaring an element type might be misleading at the API level.

We *do not* want to describe strict aliasing rules and undefined behavior in the Span API!

Reply:
I've updated this to be optional, with an explanation (hadn't noticed that a push failed on Friday.) We won't use buffer-pointer layout after all, because we do want the pointer to be raw.